haskell-cabal.el

```diff
@@ -1052,6 +1052,62 @@ source-section."
     ;; unwind
     (haskell-mode-toggle-interactive-prompt-state t)))
 
-(provide 'haskell-cabal)
 
+(defun haskell-cabal--find-tags-dir ()
+  "Return a directory where TAGS file will be generated.
+Tries to find cabal file first and if succeeds uses its location.
+If cabal file not found uses current file directory.  If current
+buffer not visiting a file returns nil."
+  (or (haskell-cabal-find-dir)
+      (when buffer-file-name
+        (file-name-directory buffer-file-name))))
+
+(defun haskell-cabal--compose-hasktags-command (&optional cabal-dir)
+  "Prepare command to execute hasktags for current file.
+By default following parameters are passed to Hasktags
+executable:
+-e - generate ETAGS file
+-x - generate additional information in CTAGS file.
+
+Tries to find cabal file location, give optional CABAL-DIR parameter to
+override it.  If cabal file not found uses current file
+directory if current buffer is visiting a file.  Otherwise returns nil.
+
+This function takes into account user's operation system: in case
+of Windows it generates simple command like
+
+hasktags --output=DIR\TAGS -x -e DIR
+
+relying on Hasktags itself to find source files;
+
+In other cases it uses `find` command to find all source files
+recursively avoiding visiting unnecessary heavy directories like
+.git, .svn, _darcs and build directories created by
+cabal-install, stack, etc."
+  (let ((dir (or cabal-dir (haskell-cabal--find-tags-dir))))
```

Comment:
This logic can go away when argument `cabal-dir` is required. One less failure scenario :)